sdk/core/azure-core/inc/http/curl/curl.hpp

```diff
@@ -77,14 +263,90 @@ namespace Azure { namespace Core { namespace Http {
 
       return curl_easy_setopt(m_pCurl, CURLOPT_WRITEDATA, (void*)this);
     }
+    // set callback for puting data into libcurl
+    CURLcode SetReadRequest()
+    {
+      auto settingUp = curl_easy_setopt(m_pCurl, CURLOPT_UPLOAD, 1L);
+      if (settingUp != CURLE_OK)
+      {
+        return settingUp;
+      }
+      settingUp = curl_easy_setopt(m_pCurl, CURLOPT_READFUNCTION, ReadBodyCallBack);
+      this->uploadedBytes = 0; // restart control counter during setup
+      if (settingUp != CURLE_OK)
+      {
+        return settingUp;
+      }
+      settingUp = curl_easy_setopt(m_pCurl, CURLOPT_READDATA, (void*)this);
+      if (settingUp != CURLE_OK)
+      {
+        return settingUp;
+      }
 
-    CURLcode Perform(Context& context, Request& request);
+      return curl_easy_setopt(m_pCurl, CURLOPT_WRITEDATA, (void*)this);
+    }
+
+    CURLcode HttpRawSend();
+    CURLcode SendBuffer(uint8_t* buffer, size_t bufferSize);
+    CURLcode ReadStatusLineAndHeadersFromRawResponse();
+    uint64_t ReadSocketToBuffer(uint8_t* buffer, size_t bufferSize);
 
   public:
-    CurlTransport();
-    ~CurlTransport();
+    CurlSession(Request& request) : m_request(request)
+    {
+      m_pCurl = curl_easy_init();
+      m_bodyStartInBuffer = 0;
+    }
 
-    std::unique_ptr<Response> Send(Context& context, Request& request) override;
+    CURLcode Perform(Context& context);
+    std::unique_ptr<Azure::Core::Http::Response> GetResponse() { return std::move(m_response); }
+    // Api for CurlStream
+    uint64_t ReadWithOffset(uint8_t* buffer, uint64_t bufferSize, uint64_t offset);
   };
 
+  class CurlTransport : public HttpTransport {
+  public:
+    std::unique_ptr<Response> Send(Context& context, Request& request) override;
+
+  }; // namespace Http
+
+  // stream to be returned inside HTTP response when using curl
+  // It keeps the ref to CurlTrasnport in order to close the handle once done
+  class CurlBodyStream : public Azure::Core::Http::BodyStream {
+  private:
+    uint64_t m_length;
+    CurlSession* m_curlSession;
+    uint64_t m_offset;
+
+  public:
+    // length comes from http response header `content-length`
+    CurlBodyStream(uint64_t length, CurlSession* curlSession)
+        : m_length(length), m_curlSession(curlSession), m_offset(0)
+    {
+    }
+
+    uint64_t Length() { return m_length; }
```

Comment:
in some places you use `this->`, in others you don't. I suggest to use `this->` everywhere possible.

Reply:
weak suggestion

Reply:
Well, I don't think this is necessary, we already have `m_` prefix for member variables.

Reply:
I am fine either way, I slightly prefer consistency, but it is not the end of the world even if we use `this->` in one place but not the other. Using `this->` is slightly technically superior - there is a case when you have global function and a template class member function named the same, and then if you call that from the template member of the template class, it is going to call the global function which is not what most expects. But is it worth to use `this->` everywhere because of this? I'd say no.

Reply:
okay